lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs

```diff
@@ -418,28 +461,27 @@ increaseZeroAdaOutputs era pp = modifyLedgerBody $
 
 -- | Internal helper to 'balanceTransaction'
 balanceTransactionWithSelectionStrategyAndNoZeroAdaAdjustment
-    :: forall era m s k ktype.
-        ( BoundedAddressLength k
-        , MonadRandom m
+    :: forall era m changeState.
+        ( MonadRandom m
         , IsRecentEra era
         )
     => Tracer m BalanceTxLog
-    -> TransactionLayer k ktype SealedTx
-    -> Maybe ([(W.TxIn, W.TxOut)] -> [CA.Script KeyHash])
-    -> Maybe ScriptTemplate
+    -> UTxOAssumptions
     -> (W.ProtocolParameters, Cardano.ProtocolParameters)
     -> TimeInterpreter (Either PastHorizonException)
     -> UTxOIndex WalletUTxO
-    -> ChangeAddressGen s
-    -> s
+    -> ChangeAddressGen changeState
+    -> changeState
     -> SelectionStrategy
     -> PartialTx era
-    -> ExceptT ErrBalanceTx m (Cardano.Tx era, s)
+    -> ExceptT ErrBalanceTx m (Cardano.Tx era, changeState)
 balanceTransactionWithSelectionStrategyAndNoZeroAdaAdjustment
     tr
-    txLayer
-    toInpScriptsM
-    mScriptTemplate
+    (UTxOAssumptions
+        txLayer
+        toInpScriptsM
+        mScriptTemplate
+        _desc)
```

Comment:
Perhaps use `NamedFieldPuns` here instead of positional pattern matching?

Reply:
Although I guess then the compiler won't warn us if we add a field to the type but not to the pattern match 🤔

I think it's fine here (no two fields of the same type) and I guess the `UTxOAssumptions` type will be revisited soon-ish anyway. Either to be converted to a sum-type or to something else.